src/operator/numpy/np_tensordot_int_axes_op-inl.h

```diff
@@ -0,0 +1,210 @@
+    return;
+  }
+
+  if (out.shape_.Size() == 0U) {
+    return;  // zero-size output, no need to launch kernel
+  }
+
+  const mxnet::TShape& a_shape = a.shape_;
+  const mxnet::TShape& b_shape = b.shape_;
+
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();  
+  CHECK_EQ(out.type_flag_, a.type_flag_)
+      << "Binary function only support input/output with the same type";
+  CHECK_EQ(out.type_flag_, b.type_flag_)
+      << "Binary function only support input/output with the same type";
+  CHECK(out.type_flag_ == kFloat32 || out.type_flag_ == kFloat64 ||
+      (out.type_flag_ == kFloat16 && ctx.run_ctx.ctx.dev_mask() == mshadow::gpu::kDevMask))
+      << "Tensordot only supports float32/float64 for CPU, and float16/float32/float64 for GPU";       
+
+  Tuple<int> a_axes_summed;
+  Tuple<int> b_axes_summed;  
+  GetSummedAxes(a_axes_summed, b_axes_summed, axes, a_shape);
+
+  Tuple<int> a_axes_remained;
+  Tuple<int> b_axes_remained;
+  Tuple<int> a_axes;
+  Tuple<int> b_axes;
+  GetReorderedAxes(a_axes_summed, a_axes_remained, a_axes, b_axes_summed, b_axes_remained, 
+    b_axes, a_shape, b_shape);
+
+  int ad1 = 1, ad2 = 1, bd1 = 1, bd2 = 1;
+  GetMatrixDimensions(ad1, ad2, bd1, bd2, a_axes_remained, a_axes_summed, 
+    b_axes_remained, b_axes_summed, a_shape, b_shape);
+
+  MSHADOW_REAL_TYPE_SWITCH(out.type_flag_, DType, {                           
+    if (a_shape.Size() == 0U || b_shape.Size() == 0U) { // 0-size input
+      if (req[0] != kAddTo) {
+        Tensor<xpu, 1, DType> out_data = out.get_with_shape<xpu, 1, DType>(
+            Shape1(out.shape_.Size()), s);
+        out_data = static_cast<DType>(0);
+      }
+      return;
+    }  
+
+    MatrixDot<xpu>(ctx, a, b, out, req[0], ad1, ad2, bd1, bd2);                         
+  });
+
+}
+
+/**
+ * forward function
+ */
+template<typename xpu>                                                  
+void TensordotIntAxesOpForward(
+    const nnvm::NodeAttrs& attrs,                     
+    const OpContext& ctx,                                
+    const std::vector<TBlob>& inputs,                 
+    const std::vector<OpReqType>& req, 
+    const std::vector<TBlob>& outputs) {                   
+
+  CHECK_EQ(inputs.size(), 2U);                                                 
+  CHECK_EQ(outputs.size(), 1U);                                               
+  CHECK_EQ(req.size(), 1U);  
+
+  const TBlob& a = inputs[0];
+  const TBlob& b = inputs[1];
+  const TBlob& out = outputs[0];  
+
+  const TensordotIntAxesParam& param = nnvm::get<TensordotIntAxesParam>(attrs.parsed);  
+  const int& axes = param.axes;   
+
+  TensordotIntAxesImpl<xpu>(axes, ctx, a, b, out, req);                                                                                                                                                  
+}     
+
+template<typename xpu>                                                       
+void TensordotIntAxesDerivativeImpl(
+    const int& axes,                       
```

Comment:
No need to pass primitive types by reference.

Reply:
Fixed. Thx.